### ISSUE TITLE
Memory leak when read messages (aka keywordInfo)

### ISSUE THREAD

**Issue by vfalduto** [googleads/google-ads-php#575]
[DO NOT INCLUDE ANY PERSONAL OR SENSITIVE DATA - MAKE SURE TO REDACT CONTENT WHEN NECESSARY]

**What versions of the client library and Google Ads API are you using?**

- Client library 8.1.0
- Google Ads API V6

**What environment are you using?**
C implementation of protobuf 
Release Date          2021-04-08 17:18:02
Release Version       3.15.8 (stable)

gRPC 
Release Date          2021-03-01 21:34:01
Release Version       1.36.0 (stable)

PHP 8.0.0
Linux c74413a656df 4.19.0-9-amd64 #1 SMP Debian 4.19.118-2+deb10u1 (2020-06-07) x86_64

**Actions taken**
```php

$query <<<QRY
      select
	ad_group_criterion.criterion_id,			
	ad_group_criterion.bid_modifier,									
	ad_group_criterion.type,
	ad_group_criterion.status,
	ad_group_criterion.negative,
	ad_group_criterion.keyword.text,
	ad_group_criterion.keyword.match_type,
        ad_group_criterion.cpc_bid_micros,
        ad_group_criterion.final_mobile_urls,
        ad_group_criterion.final_url_suffix, 
        ad_group_criterion.final_urls, 
        ad_group_criterion.url_custom_parameters,
        ad_group_criterion.tracking_url_template,
	ad_group.id,	
	campaign.id,
        campaign.bidding_strategy_type
	from ad_group_criterion
	where ad_group_criterion.type in (KEYWORD) 	
	and ad_group_criterion.negative = false
QRY;

$service = $this->client->getGoogleAdsServiceClient();
$stream = $service->searchStream($context->customerId, $query);
foreach($stream->iterateAllElements() as $googleAdsRow) {
	
        // leak memory up to 7gb
         $leak = $googleAdsRow->getAdGroupCriterion()->getKeyword();
        
        // memory around 500mb
	$noleak = json_decode($googleAdsRow->getAdGroupCriterion()->serializeToJsonString(), 1);
}
```

**Expected result**

Should not raise memory up to 7gb. (Memory_limit on php.ini was 2gb)
The current query is requesting 1.8m rows 

**Actual result**

There is a memory leak on GoogleAdsRow accessors 
Detected outside php process with htop... 

**Anything else we should know about your project / environment**

**Comment by PillowPillow** [comment]
Same issue here with a similar environment. 
Using the Protobuf C extension enabled we're observing a memory leak crashing or app during google ads querying.

No issue with the php version but we have poor performances.. not really a solution to disable the .so.

**Comment by haberman** [comment]
Taking your example:

 ```php
        // leak memory up to 7gb
         $leak = $googleAdsRow->getAdGroupCriterion()->getKeyword();
        
        // memory around 500mb
        $noleak = json_decode($googleAdsRow->getAdGroupCriterion()->serializeToJsonString(), 1);
```

One important difference between these two lines: the first one will hold onto all the memory from the entire proto (`$googleAdsRow`, or whatever parent proto it belongs to), while the second one will only hold onto the memory for the `ad_group_criterion` field.

If the entire proto is much larger than the sub-proto that you actually need to hold onto, then you'll save a lot of memory by doing a deep copy into a new message object, for example:

```php
    $criterion = new AdGroupCriterion();
    $criterion->mergeFrom($googleAdsRow->getAdGroupCriterion());
```

We could create a convenience method for making this kind of deep copy.

Note that `clone` will *not* help here, as `clone` is a shallow copy only so will continue to reference all the data from the original `$googleAdsRow`.

Does making a deep copy fix your problem?

**Comment by PillowPillow** [comment]
Hi @haberman thank your feedback !

I tried the above solution but the leak seems still there (around 2.5Gb less).

Trying to figure what's happening, we created an ultra basic sample test.php.
Even without querying the searchStream API, building an AdGroupCriterion + KeywordInfo from a json string triggers the same memory leak. 
Most of the time my process is killed during iterations.

```php
foreach(range(0, 1800000) as $i) {
      echo $i, PHP_EOL;
      $r = new AdGroupCriterion();
      $r->mergeFromJsonString('{"keyword":{"text":"foobar","matchType":"BROAD"}}', true);
      $h = $r->getKeyword();
}
```

Same issue with the GenderInfo criterion
```php
foreach(range(0, 1800000) as $i) {
      echo $i, PHP_EOL;
      $r = new AdGroupCriterion();
      $r->mergeFromJsonString('{"gender":{"type":"FEMALE"}}', true);
      $h = $r->getGender();
}
```

**Comment by vfalduto** [comment]
Hi @haberman 

I try the same script as @PillowPillow with the same result. 
We also try to set a memory_limit on my script, it's doesn't work. 

Do you need more info ? maybe a docker image ?

**Comment by haberman** [comment]
Thank you for the very compact repro. I've filed a bug against protobuf here and will investigate: https://github.com/protocolbuffers/protobuf/issues/8525

**Comment by PierrickVoulet** [comment]
The underlying Protobuf issue was fixed and released in [v3.17.1](https://github.com/protocolbuffers/protobuf/releases/tag/v3.17.1), please upgrade to this version to stop the memory leaks.
